### ISSUE TITLE
[Feature Request] Automatic Generation of Multiple Decks' Parameters

### ISSUE THREAD

**Issue by ethanbtlr** [open-spaced-repetition/fsrs4anki#314]
**Which module is related to your feature request?**
Optimizer

**Is your feature request related to a problem? Please describe.**
If I am trying to generate optimized parameters for multiple decks it takes an extreme amount of time with a large collection of decks. Due to the structure of my Anki, I have around **80 decks**. So for me to optimize each of these deck's parameters requires an inordinate amount of tedious time. I believe each deck being optimized individually will allow for better training and results.

**Describe the solution you'd like**
Given a collection (or group of decks), all decks will be optimized and have a distinct set of generated parameters. This setting could be turned on or off with a variable in the optimizer, e.g. `generate_distinct_deck_parameters = true`.

**Describe alternatives you've considered**
An alternative is doing this by hand by uploading each deck, training the model individually for every deck, and saving the parameters into the deckParams. No other automated alternatives come to my mind. 

**Additional context**
It may be redundant to generate the parameters of all parent decks. I am not sure if calculating the parent's parameters would give you any benefit. So calculating the parameters of anything other than the bottom-level decks might waste time/resources. However, I am a bit unclear on this so I wouldn't blindly follow this. I doubt it would hurt much though to generate distinct parameters for parents. You could also make this an option by setting it with variable, e.g. `generate_distinct_deck_parameters_with_parents = true`.

Here is a graphical representation of the output when  `generate_distinct_deck_parameters = true` and `generate_distinct_deck_parameters_with_parents = true` :
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/4589948/f6009600-0d6a-46bf-90bd-6310d38e770f)

Here is a graphical representation of the output when  `generate_distinct_deck_parameters = true` and `generate_distinct_deck_parameters_with_parents = false` :
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/4589948/0b2439a6-46e5-47aa-b784-fa8755e7b693)

**Comment by Expertium** [comment]
@JeffersonCorreia32a I ran statistical significance tests on your ""w" from the mother of all deck" and on ""w" from subdeck03", the difference wasn't statistically significant.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/b6cd14de-6d76-48b5-af43-29a4febcb7ac)
I'm using tests for independent samples since your parameters are for 2 different decks, not for the same deck.

That aside, I also think that running the optimizer for each deck manually is inconvenient. @L-M-Sherlock I recommend adding this to your "to do" list. Basically, either optimize parameters for all subdecks within a deck or - and this is better in my opinion - allow the user to specify more than one deck name and then run the optimizer on all decks whose names have been specified.

**Comment by JeffersonCorreia32a** [comment]
@Expertium Attached is my mother deck for any test you want to do, but I find it strange to say that parameters as different as the mother deck and subdeck03 give the same result statistically speaking, since values ​​with even smaller differences already cause a lot schedule changes.



[COLPKG E APKG JeffersonCorreia.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11895554/COLPKG.E.APKG.JeffersonCorreia.zip)

**Comment by Expertium** [comment]
While I don't think I can help with this particular issue, I appreciate you sharing your data nonetheless. I could always use more data when testing new changes to the algorithm.

**Comment by JeffersonCorreia32a** [comment]
Another very strange thing that happened in a test I did just now, doing the optimization with the collection "collection-2023-06-28@11-01-28.colpkg" the parameter w was like this:
[1.3916, 1.389, 4.8265, - 0.3405, -0.5033, 0.3054, 1.5723, -0.15, 0.9722, 2.1615, -0.0383, 0.3627, 1.1494] 

and using the mother deck "0PROJETO CONCURSO 20mil+ 28-06-2023B.apkg" the w looked like this:
[1. 2238, 1.2817, 4.8912 , -0.4236, -0.6224, 0.1222, 1.51, -0.15, 0.9093, 2.0854, -0.1139, 0.2807, 1.1038]


Since the cards contained in both are the same, shouldn't they give the same "w" value?


note: I used in these tests

timezone = 'America/Bahia'
next_day_starts_at = 12
revlog_start_date="2023-04-01"

**Comment by Expertium** [comment]
> Another very strange thing that happened in a test I did just now, doing the optimization with the collection "[collection-2023-06-28@11-01-28.colpkg](mailto:collection-2023-06-28@11-01-28.colpkg)"

I can't download it, would you send it in a different way?

**Comment by JeffersonCorreia32a** [comment]
> > Outra coisa muito estranha que aconteceu em um teste que fiz agora há pouco, fazendo a otimização com a coleção " [collection-2023-06-28@11-01-28.colpkg](mailto:collection-2023-06-28@11-01-28.colpkg) "
> 
> Não consigo fazer o download, poderia enviar de outra forma?

[coleção-2023-06-28@11-01-28.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11895998/colecao-2023-06-28%4011-01-28.zip)
[0PROJETO CONCURSO 20mil+ 28-06-2023B.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11895999/0PROJETO.CONCURSO.20mil%2B.28-06-2023B.zip)

**Comment by JeffersonCorreia32a** [comment]
@Expertium Why is it so different? shouldn't the collection reflect what's in the main deck? In April I reformulated the decks, deleting several decks and leaving little of them, and started creating new decks, what happens when we export from the anki .colpkg file? does it get data from those decks that are already turned off? is that why? In this case, which one should I use to send to the optimizer?

**Comment by JeffersonCorreia32a** [comment]
> > Why is it so different?
>
> No idea. After all, it's your collection.

Ok but what do you suggest to start from scratch? put the optimization with the APGK file instead of the COLPKG and also put the revlog_start_date=, with the date after I started using FSRS4? what was a week ago for example? revlog_start_date="2023-06-23"

Would these attitudes put an end to any data contamination that may occur?

**Comment by Expertium** [comment]
@L-M-Sherlock sorry for pinging you so often, but this is important. Check [these files](https://github.com/open-spaced-repetition/fsrs4anki/issues/314#issuecomment-1611569748).
They both should contain the same cards, and the number of reviews should be the same. But the optimizer says that one has more reviews than the other.

**Comment by user1823** [comment]
@Expertium, I am not sure how you used these files, but none of the files shared above open in my Anki or get processed in the optimizer.

@JeffersonCorreia32a, can you share the files via something like Google Drive while retaining the original file extensions?

**Comment by L-M-Sherlock** [comment]
Add a new profile and import the deck file:

<img width="754" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/010735e7-ece3-40b7-8664-5194b5ca273b">

**Comment by Expertium** [comment]
> This probably suggests that Anki doesn't clear the revlogs of cards that have been deleted.

Thta would be a **major** problem for optimization, if that's true. I hope you can figure out how so solve it.

**Comment by user1823** [comment]
> I always use:
> timezone = 'America/Bahia'
> next_day_starts_at = 12
> revlog_start_date=""

> Ok but what do you suggest to start from scratch? put the optimization with the APGK file instead of the COLPKG and also put the revlog_start_date=, with the date after I started using FSRS4? what was a week ago for example? revlog_start_date="2023-06-23"

@JeffersonCorreia32a, I suggest using the **.apkg file** for optimization instead of the .colpkg file.

Also, I suggest setting the revlog_start_date to the one on which you deleted several older decks. If you don't remember that properly, using 2023-04-01 gave me good enough results.

The loss for your deck is higher than usual. This may be because you have very small number of reviews (only 4,827⁩ reviews) or because of Anki not deleted the revlogs of the cards that you have deleted or both.

When I did the optimization for your .apkg, I got the following metrics:
```
Loss after training: 0.3417
R-squared: -5.4439
RMSE: 0.1306
```

But, FSRS still works better for you than the default Anki algorithm (SM-2). This is evident from the following metrics.

```
Universal Metric of FSRS: 0.1135
Universal Metric of SM2: 0.2045
```

The w that I got for your deck was:
[1.2238, 1.2817, 4.8912, -0.4236, -0.6224, 0.1222, 1.51, -0.15, 0.9093, 2.0854, -0.1139, 0.2807, 1.1038]

**Comment by JeffersonCorreia32a** [comment]
@user1823, thank you all for the attention given to my doubts. However, due to my questions, I ended up shifting the focus of this topic. I believe it is crucial to revisit the issue of automatically generating parameters for multiple decks. Especially considering the recent improvement in the learning rate made by @L-M-Sherlock in the optimizer, it would be great if decks could automatically receive individual parameters, especially for those with a large hierarchy of cards. @user1823 @Expertium, any ideas?

**Comment by Expertium** [comment]
@L-M-Sherlock add this issue (allow users to optimize parameters for multiple decks without having to do it manually for every deck) to your "to-do" list

**Comment by L-M-Sherlock** [comment]
> add this issue (allow users to optimize parameters for multiple decks without having to do it manually for every deck) to your "to-do" list

I'm not very interested to implement it. Because:

1. Configuring multiple set of parameters in scheduler.js is annoying. I don't recommend using too many sets of parameters.
2. It is only beneficial when the user has adequate review logs for each sub-decks. If I implement it, how to determine the threshold becomes a main problem.
3. Extracting the review logs from collection hierarchically and running loops to fit the model deck by deck require to refactor the entire optimizer code.

**Comment by Expertium** [comment]
I have an idea how to do it, it's a bit different.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/1ef6a84a-f29c-42a3-9782-06b752c6ca4f)
And at the end the user gets several `avg_w` for each deck. Basically, consecutive optimization. The user just has to upload several decks.

**Comment by Expertium** [comment]
@L-M-Sherlock what do you think of my suggestion in my comment above? Rather than optimizing subdecks within a deck, just allow the user to upload several decks, specify the necessary info for each one of them, and optimize them consecutively.

**Comment by L-M-Sherlock** [comment]
> what do you think of my suggestion in my comment above? Rather than optimizing subdecks within a deck, just allow the user to upload several decks, specify the necessary info for each one of them, and optimize them consecutively.

It requires to add more code blocks. Why not let users add code blocks by themselves?

**Comment by Expertium** [comment]
Because we want to make FSRS as user-friendly as possible, and not everyone understands code

**Comment by L-M-Sherlock** [comment]
> Because we want to make FSRS as user-friendly as possible, and not everyone understands code

But even we add those code blocks, we still need to teach the user to use them. And it would confuse those users who only want to optimize one deck.

**Comment by ilya-rusyanov** [comment]
> 
> But even we add those code blocks, we still need to teach the user to use them. And it would confuse those users who only want to optimize one deck.

If user knows how to substitute the "w" array in js script, I believe he also knows how to add code blocks...

**Comment by L-M-Sherlock** [comment]
> If user knows how to substitute the "w" array in js script, I believe he also knows how to add code blocks...

Yeah. So I think we don't need to add these code blocks by default.

**Comment by Vitorvlv** [comment]
This is now a thing when optimizing via command line, works great. Just realized that the command line is faster than doing it via collab as well, so it's going to become my main method of optimizing from now on.

**Comment by Litong18** [comment]
> This is now a thing when optimizing via command line, works great. Just realized that the command line is faster than doing it via collab as well, so it's going to become my main method of optimizing from now on.

How do you generate multiple parameters with the command line? I try doing it this morning but haven't figured out a way yet.

**Comment by galantra** [comment]
> How do you generate multiple parameters with the command line? I try doing it this morning but haven't figured out a way yet.

See https://github.com/open-spaced-repetition/fsrs-optimizer/pull/4#issuecomment-1657246182

**Comment by L-M-Sherlock** [comment]
The command line optimizer has supported this feature, so I close this issue. For details about the usage, please see: https://github.com/open-spaced-repetition/fsrs4anki/wiki/Advanced-methods-of-optimization
